src/lib/support/ChipThreadWork.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ChipThreadWork.h"
+
+#ifdef __APPLE__
+#include <dispatch/dispatch.h>
+#else
+#include <semaphore.h>
+#endif
+
+#include <platform/CHIPDeviceLayer.h>
+
+namespace chip {
+namespace ThreadWork {
+namespace {
+
+struct WorkData
+{
+    WorkCallback callback;
+#ifdef __APPLE__
```

Comment:
This part does not make any sense.  There is no "non-main" CHIP thread on darwin.  Everything is running off the work queue.  So using this `ChipMainThreadScheduleAndWait` bit there will just deadlock.

We should not be adding primitives here that assume things (especially things that are not true) about the threading model that is actually being used.

If Java code has background threads that it wants to block while dispatching things somewhere, then it should do that, but that's not a core Matter SDK facility.

Reply:
So should I remove __APPLE__ part?

on the Android platform, the matter work queue is running on a separate thread, and when we want to do sth ( stop matter app here ) called from user thread, we need this to let it run in the work queue.

Reply:
My point is that this "block current thread until this work item runs" thing should probably not be a core SDK utility, since it assumes a specific threading model.  Putting it somewhere in the Java code that wants to use it and knows that it can make the assumptions that make it safe would make more sense to me.

Reply:
changed @bzbarsky-apple